### PR TITLE
Microoptimise matching

### PR DESCRIPTION
**Pull request by redbaron**
Series of patches with minor matching optimizations.

ab65905 - adds BenchmarkQueryRange to measure effect of future changes
 e222d40 - avoids extra iteration over possible results, run matchers checks before adding to a result instead 
 1121364 - extracts logic which prepares candidate FPs into a separate method
9b15a3b - adds `fpsForLabelMatchers` and `seriesForLabelMatchers` methods and uses them where appropriate 

Before (as of ab65905):
```
go test -bench BenchmarkQueryRange -run none -benchtime=10s -benchmem
    1000	  14723524 ns/op	 5013534 B/op	   83059 allocs/op
```

After:
```
go test -bench BenchmarkQueryRange -run none -benchtime=10s -benchmem
    1000	  11553489 ns/op	 4688423 B/op	   82596 allocs/op
```

### CODE REVIEW

**Comment by redbaron** [comment]
rebased

**Comment by redbaron** [comment]
Can someone have a look please? It is up to 30% speedup on QueryRange benchmark

**Comment by beorn7** [comment]
It's on my review queue, I'm just totally overloaded, as usual...

File: storage/local/storage.go [hunk 1]
```diff
@@ -621,29 +621,104 @@ func (s *MemorySeriesStorage) metricsForLabelMatchers(
...
+func (s *MemorySeriesStorage) fpsForLabelMatchers(
+	from, through model.Time,
+	matchers ...*metric.LabelMatcher,
+) (map[model.Fingerprint]struct{}, error) {
+	candidateFPs, matchersToCheck, err := s.candidateFPsForLabelMatchers(matchers...)
+	if err != nil {
+		return nil, err
+	}
+
+FP_LOOP:
```
Comment (beorn7) [inline storage/local/storage.go]: As above.

File: storage/local/storage.go [hunk 2]
```diff
@@ -621,29 +621,104 @@ func (s *MemorySeriesStorage) metricsForLabelMatchers(
...
+	matchers ...*metric.LabelMatcher,
+) (map[model.Fingerprint]metric.Metric, error) {
+
+	candidateFPs, matchersToCheck, err := s.candidateFPsForLabelMatchers(matchers...)
+	if err != nil {
+		return nil, err
+	}
+
+	result := map[model.Fingerprint]metric.Metric{}
+FP_LOOP:
```
Comment (beorn7) [inline storage/local/storage.go]: As above.

File: storage/local/storage.go [hunk 3]
```diff
@@ -888,40 +963,42 @@ func (s *MemorySeriesStorage) seriesForRange(
...
-	s.fpLocker.Lock(fp)
-	defer s.fpLocker.Unlock(fp)
-
-	series := s.seriesForRange(fp, from, through)
+	fp, series := fpsPair.fp, fpsPair.series
 	if series == nil {
 		return nopIter
 	}
+
+	s.fpLocker.Lock(fp)
```
Comment (beorn7) [inline storage/local/storage.go]: As discussed in the other PR, the performance gain of not going with `defer` is probably negligible in this context, and will go away with Go1.8 anyway. Better use the cleaner code here (i.e. unlock with defer as before).

File: storage/local/storage.go [hunk 4]
```diff
@@ -888,40 +963,42 @@ func (s *MemorySeriesStorage) seriesForRange(
...
```
Comment (beorn7) [inline storage/local/storage.go]: Same as above, unlock with defer.

File: storage/local/storage.go [hunk 5]
```diff
@@ -888,40 +963,42 @@ func (s *MemorySeriesStorage) seriesForRange(
 }
 
 func (s *MemorySeriesStorage) preloadChunksForRange(
-	fp model.Fingerprint,
+	fpsPair fingerprintSeriesPair,
```
Comment (beorn7) [inline storage/local/storage.go]: `fpsPair` is misleading. (Pair of two FPs?). I'd just call it `pair`.

File: storage/local/storage.go [hunk 6]
```diff
@@ -888,40 +963,42 @@ func (s *MemorySeriesStorage) seriesForRange(
...
+		s.fpLocker.Unlock(fp)
 		return nopIter
 	}
+	s.fpLocker.Unlock(fp)
 	return iter
 }
 
 func (s *MemorySeriesStorage) preloadChunksForInstant(
-	fp model.Fingerprint,
+	fpsPair fingerprintSeriesPair,
```
Comment (beorn7) [inline storage/local/storage.go]: See above.

File: storage/local/test_helpers.go [hunk 7]
```diff
@@ -65,3 +65,7 @@ func NewTestStorage(t testutil.T, encoding chunkEncoding) (*MemorySeriesStorage,
 
 	return storage, closer
 }
+
+func mkFpSeriesPair(s *MemorySeriesStorage, fp model.Fingerprint) fingerprintSeriesPair {
```
Comment (beorn7) [inline storage/local/test_helpers.go]: Capitalization nit: `mkFPSeriesPair`. But given the name of the returned type (where FP is not abbreviated), and given that "make" is rarle abrreviated "mk" in typical Go constructors, why not call it `makeFingerprintSeriesPair` after all?

**Comment by beorn7** [comment]
@redbaron Ping?

**Comment by redbaron** [comment]
@beorn7 will be able to do sometime this week

**Comment by redbaron** [comment]
@beorn7 , made suggested changes

**Comment by redbaron** [comment]
sorry, I didn't  know this PR is holding a release back.  I'll keep this tab open and expect to be able to make fixes if needed within an hour.

File: storage/local/storage.go (hunk 1 above)
Comment (beorn7) [inline storage/local/storage.go]: With my comment above, this comment finally makes sense. Use `FPLoop:`

File: storage/local/storage.go (hunk 2 above)
Comment (beorn7) [inline storage/local/storage.go]: With my comment above, this comment finally makes sense. Use FPLoop:

**Comment by beorn7** [comment]
Thanks. The only missing nit is the one about the loop labels. See comments above. After that, we can merge.

File: storage/local/storage.go [hunk 8]
```diff
@@ -621,29 +621,104 @@ func (s *MemorySeriesStorage) metricsForLabelMatchers(
...
+	from, through model.Time,
+	matchers ...*metric.LabelMatcher,
+) ([]fingerprintSeriesPair, error) {
+	candidateFPs, matchersToCheck, err := s.candidateFPsForLabelMatchers(matchers...)
+	if err != nil {
+		return nil, err
+	}
+
+	result := []fingerprintSeriesPair{}
+FP_LOOP:
```
Comment (redbaron) [inline storage/local/storage.go]: done

**Comment by beorn7** [comment]
Sorry for the nit, but ALLCAPS is also not very Go idiomatic. Please use `FPLoop` as label name.

**Comment by redbaron** [comment]
fixed, I wasn't reading careful enough :)

**Comment by beorn7** [comment]
Thank you. Merging now.

**Comment by redbaron** [comment]
Do you still keep authors.md up to date?

**Comment by beorn7** [comment]
In principle, yes. It definitely needs a major update. I had a script for it somewhere…

**Pull request merged**